fix: Unable to view details of product after scanning

bhattabhi013 opened the pull request:
### What
Fixes Unable to view details of the product after scanning.

### Screenshot
<img width="270" alt="image" src="https://user-images.githubusercontent.com/47862474/166119236-ba49a705-7f1e-4130-b6fb-d4068537a3cb.png">

### Fixes bug(s)
- #1725 


### Part of 
- https://github.com/openfoodfacts/smooth-app/issues/525
(please be as granular as possible)

bhattabhi013 committed:
fix: Unable to view details of product after scanning

bhattabhi013 committed:
moved the provider to main.dart file

monsieurtanuki reviewed (commented):
I have no time to double-check now, and I'll have no time tomorrow (Sunday May 1st).
OK it looks fair to add ContinuousScanModel in main.dart, but I would be VERY surprised if you don't impact ALL the other places in the code that called/created an instance of `ContinuousScanModel`. It's like a singleton. But if it's also created somewhere else, that's a call for confusion and bugs.

bhattabhi013 committed:
cross button changes

bhattabhi013 commented:
Hi @monsieurtanuki,
I've moved the `ContinuousScanModel` to `main.dart` and added `if (isRemovable && !isSelectable)` to `product_title_card.dart` to remove cross when SummaryCard is in full screen.

SummaryCard in full screen(cross not required)
<img width="266" alt="image" src="https://user-images.githubusercontent.com/47862474/166158909-b1b2f2b2-5ef4-4c1f-91ed-76d4919d921f.png">

SummaryCard not in full screen(cross required to remove from carousel)
<img width="269" alt="image" src="https://user-images.githubusercontent.com/47862474/166158896-ac69a45f-b2c8-4f3c-a94e-be6ca35a833b.png">


I also checked the impact at onboarding which caused #1720 and everything works fine there.

monsieurtanuki reviewed (changes requested):
Hi @bhattabhi013!
I have no strong opinion about when to put the cross, except that we should check first if the product is in the scan carousel if the idea is to put a button to remove the product from the carousel.
But, my main concern is that we should call only once in the whole app `await ContinuousScanModel().load(localDatabase)`. Please fix that first.

bhattabhi013 commented:
Hi @monsieurtanuki,
I apologize as I'm unable to understand what you are trying to convey but I understood that there should be only one call of `await ContinuousScanModel().load(localDatabase)` and I can confirm that it is called only once i.e. in `scan_page`. Please let me know if I'm missing something.

M123-dev commented:
I think what he means it that we now have two Providers of type ContinuousScanModel but only one should be in the whole widget tree

monsieurtanuki commented:
@bhattabhi013 `ContinuousScanModel()` must be called only once in the app, in `main.dart`. And the value can be retrieved in the whole app through a provider.

bhattabhi013 committed:
removed extra instances of ContinuousScanModel()

monsieurtanuki commented on packages/smooth_app/lib/helpers/picture_capture_helper.dart:
```diff
@@ -41,7 +41,8 @@ Future<bool> uploadCapturedPicture(
     BuildContext context, String barcode) async {
-  final ContinuousScanModel? model =
-      await ContinuousScanModel().load(context.read<LocalDatabase>());
+  final ContinuousScanModel? model = await context
+      .read<ContinuousScanModel>()
+      .load(context.read<LocalDatabase>());
   await model?.onCreateProduct(barcode);
```
```suggestion
  final ContinuousScanModel model = context.read<ContinuousScanModel>();
  await model.onCreateProduct(barcode);
```

bhattabhi013 replied:
Changes done

monsieurtanuki commented on packages/smooth_app/lib/main.dart:
```diff
@@ -110,6 +111,8 @@ class _SmoothAppState extends State<SmoothApp> {
   final UserManagementProvider _userManagementProvider =
       UserManagementProvider();
 
+  final ContinuousScanModel _continuousScanModel = ContinuousScanModel();
```
Please load it too in the async init.

bhattabhi013 replied:
added `ContinuousScanModel` at init.

monsieurtanuki commented on packages/smooth_app/lib/pages/scan/scan_page.dart:
```diff
@@ -35,9 +35,10 @@ class _ScanPageState extends State<ScanPage> {
-    final LocalDatabase localDatabase = context.watch<LocalDatabase>();
     if (_model == null) {
-      _model = await ContinuousScanModel().load(localDatabase);
+      _model = await context
+          .read<ContinuousScanModel>()
+          .load(context.read<LocalDatabase>());
```
```suggestion
      _model = await context.read<ContinuousScanModel>();
```

bhattabhi013 replied:
Changes done.

monsieurtanuki reviewed (commented):
Hi @bhattabhi013, it looks more like that indeed!
Still, I think we could `load` the model at init time too, right?
Please have a look at my comments and suggestions.

bhattabhi013 committed:
added provider to async init

bhattabhi013 committed:
loaded at init

monsieurtanuki commented on packages/smooth_app/lib/pages/scan/scan_page.dart:
```diff
@@ -35,9 +34,8 @@ class _ScanPageState extends State<ScanPage> {
-      _model = await ContinuousScanModel().load(localDatabase);
+      _model = context.read<ContinuousScanModel>();
     } else {
       await _model?.refresh();
```
```suggestion
      await _model!.refresh();
```

bhattabhi013 replied:
Changes done.

monsieurtanuki commented on packages/smooth_app/lib/main.dart:
```diff
@@ -64,6 +65,7 @@ late UserPreferences _userPreferences;
 late ProductPreferences _productPreferences;
 late LocalDatabase _localDatabase;
 late ThemeProvider _themeProvider;
+late ContinuousScanModel _continuousScanModel;
```
```suggestion
final ContinuousScanModel _continuousScanModel = ContinuousScanModel();
```

bhattabhi013 replied:
This looks like a better approach, changes are done.

monsieurtanuki commented on packages/smooth_app/lib/main.dart:
```diff
@@ -84,6 +86,7 @@ Future<bool> _init1() async {
   );
   _userPreferences = await UserPreferences.getUserPreferences();
   _localDatabase = await LocalDatabase.getLocalDatabase();
+  _continuousScanModel = (await ContinuousScanModel().load(_localDatabase))!;
```
```suggestion
  await _continuousScanModel.load(_localDatabase);
```

bhattabhi013 replied:
Changes done.

monsieurtanuki reviewed (approved):
Thank you @bhattabhi013!
I just have minor suggestions; feel free to ignore them.

bhattabhi013 committed:
improvization

monsieurtanuki reviewed (approved):
@bhattabhi013 Let's try it that way!
Possibly, your changes may have a bad impact on the screenshot generation, but that's another story!

bhattabhi013 commented:
Thanks, @monsieurtanuki,
I'm really thankful for all your support and guidance during this PR, I was able to get a better understanding of the provider and app code. 
On a side note, I feel this may not impact the screenshot generation because we are using 2 different `init's` in `main.dart` .

monsieurtanuki commented:
> On a side note, I feel this may not impact the screenshot generation because we are using 2 different `init's` in `main.dart`

I don't think it's appropriate to use logic regarding screenshot generation. Just painful experiments do the trick. Believe me!
Actually we're using 2 `init`s just because it seemed to work that way. Don't ask me why.
